index/upside_down/reader.go

```diff
@@ -104,11 +105,23 @@ func (r *UpsideDownCouchTermFieldReader) Next(preAlloced *index.TermFieldDoc) (*
 
 func (r *UpsideDownCouchTermFieldReader) Advance(docID index.IndexInternalID, preAlloced *index.TermFieldDoc) (*index.TermFieldDoc, error) {
 	if r.iterator != nil {
-		tfr := NewTermFrequencyRow(r.term, r.field, docID, 0, 0)
-		r.iterator.Seek(tfr.Key())
+		if r.tfrNext == nil {
+			r.tfrNext = &TermFrequencyRow{}
+		}
+		tfr := InitTermFrequencyRow(r.tfrNext, r.term, r.field, docID, 0, 0)
+		keySize := tfr.KeySize()
+		if cap(r.keyBuf) < keySize {
+			r.keyBuf = make([]byte, keySize)
+		}
```

Comment:
I will +2 this, but I think the code is getting way messier than it needs, and its all these ad-hoc perf optimizations.  Here is what I see:

1.  We create an object
2.  We use the object to create a []byte
3.  We add code to cache/reuse the object
4.  We add code to cache/reuse the []bte

When in fact:
1.  We never needed the object in the first place.  We should just have a method that takes the args and goes directly to the bytes
2.  That method could take in the existing []byte, and the optional reuse/grow buffer logic should be hidden inside that method (returns []byte which might be reused or new larger one)

I think changes like this would keep the TermFieldReader code easier to read.